ls1intum/Artemis
# [Exam] File upload exercises for exam mode

### Checklist

### Motivation and Context
File upload exercise CRUD operations must be adapted for exam mode.

### Description
Client:
- [x] Created new client routes for text-exercise update and creation
For creation: ':examId/exercise-groups/:groupId/file-upload-exercises/new'
For update: ':examId/exercise-groups/:groupId/file-upload-exercises/:exerciseId/edit'
- [x] Load file upload exercise module
- [x] Extend FileUploadExerciseResolver by fetching of exerciseGroup for the new routes
- [x] Deactivate categories in exam mode, as this relies on a Course
- [x] Added "Add file upload exercise" button to exercise group component 
- [x] Refactored the TextExerciseRowButtonsComponent. This component is responsible for displaying  edit and delete button of exam exercises. The component can now be used for all exercise types.
- [x] Added exam mode to TextExercise and FileUploadExercise Detail Component. 

Server:
- [x] Deactivate Tutor/Student notification on exercise creation/update
- [x] Add additional authorization check to GET file upload exercise route. Only instructors and admins should be allowed to access an exam file upload exercise
- [x] GET for exam text and file upload exercises now also returns the exam and course. The client can use this to build the right routes for example for the Edit button of the Detail Component
- [x] Adapted authorization check for deletion
- [x] Refactored course retrieval over ExerciseGroups

**We still have a problem with client route consistency. For example we only use the exerciseId of edit routes such as `course-management/:courseId/exams/:examId/exercise-groups/:groupId/file-upload-exercises/:exerciseId/edit` to retrieve the exercise. The other id are not used and could be anything. This is not a security problem as the server still does all the authorization checks. Nevertheless this should be fixed in the future**

### Steps for Testing
_Ensure that the existing CRUD operations for regular file upload exercises still work_

1. Log in as admin or instructor
2. Create a regular file upload exercise
--> Exercise was created
3. Update the exercise
--> Exercise was updated
4. View the exercise
--> Can open the page
5. Delete the exercise
--> Exercise was deleted

_Ensure that CRUD operations for exam file upload exercises work_

1. Log in as admin or instructor
2. Create an exam
3. Create an exercise group
4. Create a file upload exercise
--> Exercise was created
5. Edit the file upload exercise
--> Can update the file upload exercise
6. Navigate to `/course-management/:courseId/file-upload-exercises/:exerciseId` using the id of an exam file upload exercise
--> Should be able to see the exercise
7. Delete the exercise
--> Exercise was deleted

_Ensure correct authorization for file upload exercises_
1. Log in as a tutor or student
2. Navigate to `course-management/:courseId/exams/:examId/exercise-groups/:groupId/file-upload-exercises/new`
--> You should not be able to access the route
4. Use the id of an exam exercise for the following steps
5. Navigate to `course-management/:courseId/exams/:examId/exercise-groups/:groupId/file-upload-exercises/:exerciseId/edit`
--> You should not be able to access the route
7. Navigate to `/course-management/:courseId/file-upload-exercises/:exerciseId`
--> You should not be able to see the exercise

Please also test regular and exam text exercises again as the component which displays buttons such as edit, view, teams (for regular exercise) was refactored.

### Screenshots
![fileupload](https://user-images.githubusercontent.com/16407766/84939602-e0dc8c80-b0de-11ea-9bdd-7f8234e23e49.png)
